### PR TITLE
fix: UHF-10736: Try to make sure that uploaded attachments are properly set to application data.

### PR DESCRIPTION
**Pull request by jiisuominen**
# [UHF-10736](https://helsinkisolutionoffice.atlassian.net/browse/UHF-10736)

Application attachments still has some issues with the UI missing integrationID in some cases even though all navigation should be disabled. 

This PR tries to fix this in 2 prone approach:
- Add an overlay to the page when ajax is processing, this should prevent any navigations or interruptions from user to ajax process.
- Since the file uploads itself have always been working, we go through attachments from ATV document and if there's a matching filename added to some attachment-field on application we add integrationID to it so that it can be downloaded to Avus2 via integration.

## What was done

* added overlay for ajax processing
* refactored integrationID fixing code from resend application functionality to custom service
* attachments are checked on every save, whether save is done to ATV or via integration. 
* small fix to uploading process so that the same object is used through the whole process. This makes saving of applications more robust.
* fixed tests for application 29, there was some fields missing.

## How to install

* Make sure your instance is up and running on correct branch.
  * `git checkout feature/UHF-10736`
  * `make fresh`
* Run `make drush-cr`

## How to test

* [x] Create new application which has lot of attachment fields, eg [application 29](https://hel-fi-drupal-grant-applications.docker.so/fi/tietoa-avustuksista/muut-avustukset/kaupunginhallitus-yleisavustus)
* [x] Fill other pages but attachments, save as DRAFT
* [x] Open GrantsAttacments file in your IDE, and comment out integrationID part of the upload process @ [line 723 or so](https://github.com/City-of-Helsinki/hel-fi-drupal-grants/blob/6bb3fd69bb22a45a9464b6177886daeefabcbe47/public/modules/custom/grants_attachments/src/Element/GrantsAttachments.php#L723)
* [x] Add attachments to all fields
* [x] Save again as a DRAFT
* [x] Open application data from ATV with postman or PhpStorm http queries (should be found under tools -folder). You can also check the application state from [ResendApplication](https://hel-fi-drupal-grant-applications.docker.so/fi/admin/tools/resend-applications) functionality, just insert application number and click getStatus
* [x] See that the integrationID's are present even if they are omitted from form data
* [ ] Send application to Avus2, make sure from Tero that the attachments are actually downloadable in Avus2
* [ ] Check that code follows our standards

## Designers review

* [x] This PR does not need designers review
* [ ] This PR has been visually reviewed by a designer (Name of the designer)

## Automatic- / Regression tests

* [ ] This PR makes no changes that effects any tests. (This will be caught in automatic testing later on, but please, please run regression tests always on PR before asking for a review)
* [x] This PR passes regression tests. (`make test-pw`)

**This has no e2e tests written for these integrationID errors, should we?**

[UHF-10736]: https://helsinkisolutionoffice.atlassian.net/browse/UHF-10736?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: public/modules/custom/grants_attachments/src/AttachmentFixerService.php [hunk 1]
```diff
@@ -0,0 +1,251 @@
...
+ */
+final class AttachmentFixerService {
+
+  use StringTranslationTrait;
+
+  /**
+   * Constructs an AttachmentFixerService object.
+   */
+  public function __construct(
+    private readonly Messenger $messenger,
```
Comment (tuutti) [inline public/modules/custom/grants_attachments/src/AttachmentFixerService.php:22]: Messenger service is not used and can be removed

File: public/modules/custom/grants_attachments/src/AttachmentFixerService.php [hunk 2]
```diff
@@ -0,0 +1,251 @@
...
+   *   Array of booleans.
+   */
+  public function areAttachmentsOk(
+    mixed $events,
+    mixed $attachment,
+    mixed $attachmentInfo,
+    mixed $appEnv,
+  ): array {
+    // Look for events from us, the handler.
+    $handlerOk = $this->filterEventsByTypeAndFilename($events, 'HANDLER_ATT_OK', $attachment['filename']);
```
Comment (tuutti) [inline public/modules/custom/grants_attachments/src/AttachmentFixerService.php:143]: `filterEventsByTypeAndFilename` expects `$events` and `$attachment` to be an array, but this method allows `mixed` type for both. The type should be checked/enforced before calling the method.

File: public/modules/custom/grants_attachments/src/AttachmentFixerService.php [hunk 3]
```diff
@@ -0,0 +1,251 @@
...
+    // Look for events from AVUSTUS2.
+    $avus2Ok = $this->filterEventsByTypeAndFilename($events, 'AVUSTUS2_ATT_OK', $attachment['filename']);
```
Comment (tuutti) [inline public/modules/custom/grants_attachments/src/AttachmentFixerService.php:145]: Same as above.

File: public/modules/custom/grants_attachments/src/AttachmentFixerService.php [hunk 4]
```diff
@@ -0,0 +1,251 @@
...
+    // Look for errors from AVUSTUS2.
+    $avus2Error = $this->filterEventsByTypeAndFilename($events, 'AVUSTUS2_ATT_ERROR', $attachment['filename']);
```
Comment (tuutti) [inline public/modules/custom/grants_attachments/src/AttachmentFixerService.php:147]: Same as above.

File: public/modules/custom/grants_attachments/src/AttachmentFixerService.php [hunk 5]
```diff
@@ -0,0 +1,251 @@
...
+  private function checkAttachmentInfo(mixed $attachmentInfo, mixed $attachment, mixed $appEnv): bool {
+    // If no attachments, nothing to check.
+    if (!is_array($attachmentInfo)) {
+      return FALSE;
+    }
+
+    // Loop attachment info and check if we have the attachment added.
+    foreach ($attachmentInfo as $info) {
+      // Check if filename and integrationID are found.
+      $filenameFound = $this->findValueById($info, 'fileName', $attachment['filename']);
```
Comment (tuutti) [inline public/modules/custom/grants_attachments/src/AttachmentFixerService.php:200]: `findValueById` expects `$info` and `$attachment` to be an array, but this method allows `mixed` type for both. The type should be enforced before calling the method.

**Review by tuutti — CHANGES_REQUESTED** [review]

**Review by tuutti — APPROVED** [review]

**Review by juho-lehmonen — APPROVED** [review]

**Pull request merged**